File: 2024/December/SherlockAndAnagrams.java
// Time: O(26 * n^2)  Space : O(26 + n^2)  
public static int sherlockAndAnagrams(String s) {
    int ans = 0;
    HashMap<String, Integer> hm = new HashMap<>();

    // Generate all substrings
    for (int i = 0; i < s.length(); i++) {
        int[] charCount = new int[26]; // Frequency array for characters
        for (int j = i; j < s.length(); j++) {
            // Update frequency for the current character
            charCount[s.charAt(j) - 'a']++;

            // Convert frequency array to a unique string representation
            String signature = Arrays.toString(charCount);

            // Update the HashMap
            hm.put(signature, hm.getOrDefault(signature, 0) + 1);
        }
    }

    // Calculate the number of anagram pairs
    for (String key : hm.keySet()) {
        int val = hm.get(key);
        if (val > 1) {
            ans += (val * (val - 1)) / 2; // Combination formula
        }
    }

    return ans;
}

File: 2025/January/SherlockAndAnagrams.ts
// Time & Space : O(n^2)
function sherlockAndAnagrams(s: string): number {
    let signatureMap = new Map<string, number>();
    let ans = 0;

    // Generate all substrings
    for (let i = 0; i < s.length; i++) {
        let substrMap = new Array(26).fill(0); // Initialize with zeros

        for (let j = i; j < s.length; j++) {
            substrMap[s.charCodeAt(j) - 'a'.charCodeAt(0)] += 1; // Update frequency count

            const signature = JSON.stringify(substrMap); // Convert array to string as a unique key
            if (!signatureMap.has(signature)) {
                signatureMap.set(signature, 0);
            }
            signatureMap.set(signature, signatureMap.get(signature) + 1);
        }
    }

    // Count anagrammatic pairs
    for (let val of signatureMap.values()) {
        if (val > 1) {
            ans += (val * (val - 1)) / 2; // Combination formula
        }
    }

    return ans;
}
